src/io/flutter/utils/AndroidUtils.java

```diff
@@ -114,4 +185,451 @@ private static String validateName(String name, boolean requiresSeparator) {
+      COEDIT_TRANSFORMED_PROJECTS.add(project);
+    }
+  }
+
+  private static boolean isCoeditTransformedProject(@NotNull Project project) {
+    if (project.isDisposed()) {
+      return true;
+    }
+    Iterator<Project> iter = COEDIT_TRANSFORMED_PROJECTS.iterator();
+    //noinspection WhileLoopReplaceableByForEach
+    while (iter.hasNext()) { // See WeakList.iterator().
+      Project p = iter.next();
+      if (project.equals(p) && !p.isDisposed()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // Copied from org.jetbrains.plugins.gradle.execution.GradleRunAnythingProvider.
+  @NotNull
+  @SuppressWarnings("DuplicatedCode")
+  private static Map<ProjectData, MultiMap<String, String>> getTasksMap(Project project) {
+    Map<ProjectData, MultiMap<String, String>> tasks = new LinkedHashMap<>();
+    for (GradleProjectSettings setting : GradleSettings.getInstance(project).getLinkedProjectsSettings()) {
+      final ExternalProjectInfo projectData =
+        ProjectDataManager.getInstance().getExternalProjectData(project, GradleConstants.SYSTEM_ID, setting.getExternalProjectPath());
+      if (projectData == null || projectData.getExternalProjectStructure() == null) continue;
+
+      MultiMap<String, String> projectTasks = MultiMap.createOrderedSet();
+      for (DataNode<ModuleData> moduleDataNode : getChildren(projectData.getExternalProjectStructure(), ProjectKeys.MODULE)) {
+        String gradlePath;
+        String moduleId = moduleDataNode.getData().getId();
+        if (moduleId.charAt(0) != ':') {
+          int colonIndex = moduleId.indexOf(':');
+          gradlePath = colonIndex > 0 ? moduleId.substring(colonIndex) : ":";
+        }
+        else {
+          gradlePath = moduleId;
+        }
+        for (DataNode<TaskData> node : getChildren(moduleDataNode, ProjectKeys.TASK)) {
+          TaskData taskData = node.getData();
+          String taskName = taskData.getName();
+          if (isNotEmpty(taskName)) {
+            String taskPathPrefix = ":".equals(gradlePath) || taskName.startsWith(gradlePath) ? "" : (gradlePath + ':');
+            projectTasks.putValue(taskPathPrefix, taskName);
+          }
+        }
+      }
+      tasks.put(projectData.getExternalProjectStructure().getData(), projectTasks);
+    }
+    return tasks;
+  }
+
+  private static PsiElement findInclude(GradleSettingsFile parsedSettings, String name) {
+    Map<String, GradleDslElement> elements = parsedSettings.getPropertyElements();
+    GradleDslElement includes = elements.get("include");
+    if (includes == null) {
+      return null;
+    }
+    for (GradleDslElement include : includes.getChildren()) {
+      PsiElement expr = ((GradleDslLiteral)include).getExpression();
+      if (expr == null) {
+        continue;
+      }
+      String text = expr.getText();
+      if (name.equals(text.substring(2, text.length() - 1))) {
+        return expr;
+      }
+    }
+    return null;
+  }
+
+  private static class CoEditHelper {
```

Comment:
My understanding is that the logic from here to the end of the file is focused on rewriting files and is something that we might want to share with the VS Code plugin, which is where we would write code in Dart as we discussed yesterday.

The places that dependencies couple this with the rest of the file are:
* `parseSettings()`
* `findInclude()` uses the Psi tree.

We'll need to understand the requirements for a VS Code application of the logic to identify how to prune, or if it's worth doing.

Reply:
Correct. We'd also want to move the functionality of `isVanillaAddToApp()` into the external service. Once we've done that there won't be any dependencies left, since both `parseSettings()` and `findInclude()` are only used by `isVanillaAddToApp()` and `enableCoediting()`.

Reply:
Code using the Psi tree could be done use the analyzer apis in a Dart tool.